### DISCUSSION TITLE
autocompleting tags and other small QOL things

### DISCUSSION THREAD

**Discussion by astholkohtz** [daedsidog/evedel#7] [Ideas]
Hi, thank you for this life-changing package. 

Just a couple ideas that I think would be great to have. Not a feature request, do what you think is best!

1) autocompletion when adding tag queries. I always forget the exact tags I have given to my references
2) only one command for tags that modifies tags or tag queries depending on whether we are inside a ref or a directive
3) functions like inside-reference-p and inside-directive-p (should be self explanatory)
4) one toggle to tell the current query to just use entire buffers as context (including the current one), for lazy people working on small files.
5) a navigation menu like consult-imenu. 

and most importantly, diff, just like gptel does with gptel rewrite. maybe it's possible to standardize the outputs and treat them exactly as the overlays that gptel-rewrite generates, or just use and integrate with gptel-rewrite altogether?

**Comment by daedsidog** [comment]
Thanks for using the package. I appreciate the feedback.

1. This is planned
2. I like this idea
3. I could add these. Could you tell me what you would like to use them for, though? I'm curious.
4. Couldn't you use toplevel references for this sort of thing? Though regardless, there are many permutations for options of what context to use, I'll think of a better solution rather than having a binary custom variable for each one like there is now.
5. This is sort of in the works as part of better integration with gptel, I'm still not sure how to approach this so I left it last for now.

Diff is probably the highest priority.

**Comment by astholkohtz** [comment]
1. great
2. fantastic
3. just for users that like situational behavior. Might be useful to implement stuff like "go to next ref" or "go to next dir". This kind of fine-grained behavior is a bit too personal to deserve being in the source, but it's nice to have little helper functions that might facilitate implementing this sort of thing. Initially, I wanted to have those so I could have implemented 2 myself without bothering you.

4. I don't understand this:
is:bufferlevel: Returns only references that contain the entire buffer.
does it mean that I have to select all the buffer and make it into a reference?
The way I achieve this is by making the whole buffer a reference, and then leaving this option true:
 (customize-set-variable 'evedel-empty-tag-query-matches-all t)
But if I do this, then the whole buffer is unnecessarily highlighted as a reference, which is overkill, because I know I'm using the whole buffer and don't need the visual cue.

I think the following workflow would be great:

  * don't define any references, and have all the directives get the whole buffer as context. Morally, this is a bit against the philosophy of the package, and dangerously similar to what gptel does in the first place, but in practice, when you start working on a project and just have a couple hundred lines, it's still practical to just send the whole buffer. So, you'd just put down a directive, activate a query, i.e., 'entire-buffer', and start working. As the code grows, then you start actually defining references to point the LLM to the right direction and save tokens. Now that I describe this workflow, I realize that it would probably be best to have a global toggle to either make all directives use the whole buffer, or to make them use references instead (the current behavior).

5. very nice!

Re: diff: that'd make the whole thing a real pleasure to use!

**Reply by daedsidog** [reply]
>  Might be useful to implement stuff like "go to next ref" or "go to next dir". This kind of fine-grained behavior is a bit too personal to deserve being in the source,

These already exist, by the way. See [navigation](https://github.com/daedsidog/evedel?tab=readme-ov-file#navigation).

> But if I do this, then the whole buffer is unnecessarily highlighted as a reference, which is overkill, because I know I'm using the whole buffer and don't need the visual cue.

Buffer-level references are a special case and shouldn't visually highlight the entire buffer. If they do that for you, it's a bug.

I'm less keen on investing in this domain anyway, since I believe once there is a form of RAG (whenever that will be...), it should be able to automatically partition all your references and use the necessary parts, without fiddling with option permutations.

**Reply by astholkohtz** [reply]
you're right, buffer-level references are indeed not highighted. that's neat! In that case, I think it's all good and I agree with you and number 4 I had suggested is unnecessary.
